File: backend/src/main/java/org/icpclive/events/ContestInfo.java
package org.icpclive.events;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.TreeSet;
import java.util.concurrent.BlockingQueue;
import java.util.stream.Stream;

public abstract class ContestInfo {
    public int teamNumber;
    public int problemNumber = 0;
    private long startTime = 0;
    public List<ProblemInfo> problems = new ArrayList<>();
    public long lastTime;
    private static String[] hashtags;

    public enum Status {
        BEFORE,
        RUNNING,
        PAUSED,
        OVER
    }

    public Status status = Status.BEFORE;

    public static int CONTEST_LENGTH = 5 * 60 * 60 * 1000;
    public static int FREEZE_TIME = 4 * 60 * 60 * 1000;
    public static final TreeSet<String> GROUPS = new TreeSet<>();

    protected ContestInfo() {
    }

    protected ContestInfo(int problemNumber) {
        this.problemNumber = problemNumber;
    }

    public int getTeamsNumber() {
        return teamNumber;
    }

    public int getProblemsNumber() {
        return problemNumber;
    }

    public void setStatus(Status status) {
        System.err.println("New status: " + status);
        lastTime = getCurrentTime();
        this.status = status;
    }

    public long getStartTime() {
        return startTime;
    }

    public void setStartTime(long startTime) {
        System.err.println("Set start time " + new Date(startTime));
        this.startTime = startTime;
    }

    public long getTimeFromStart() {
//        if (status == Status.BEFORE) {
//            return 0;
//        }
        return (long) ((System.currentTimeMillis() - startTime) * EventsLoader.getInstance().getEmulationSpeed());
    }

    public long getCurrentTime() {
        return getCurrentTime(0);
    }

    public long getCurrentTime(int delay) {
        switch (status) {
            case BEFORE:
                return 0;
            case PAUSED:
                return lastTime;
            case RUNNING:
                return startTime == 0 ? 0 :
                        Math.min(getTimeFromStart(), ContestInfo.CONTEST_LENGTH);
            case OVER:
                if (delay != 0) {
                    return Math.min(getTimeFromStart(), ContestInfo.CONTEST_LENGTH + delay);
                }
                return ContestInfo.CONTEST_LENGTH;
            default:
                return 0;
        }
    }

    public boolean isFrozen() {
        return getCurrentTime() >= ContestInfo.FREEZE_TIME;
    }

    public abstract TeamInfo getParticipant(String name);

    public abstract TeamInfo getParticipant(int id);

    public abstract TeamInfo getParticipantByHashTag(String hashTag);

    public abstract TeamInfo[] getStandings();

    public abstract TeamInfo[] getStandings(OptimismLevel optimismLevel);

    public static final String ALL_REGIONS = "all";

    public TeamInfo[] getStandings(String group, OptimismLevel optimismLevel) {
        if (ALL_REGIONS.equals(group)) {
            return getStandings(optimismLevel);
        }
        TeamInfo[] infos = getStandings(optimismLevel);
        return Stream.of(infos).filter(x -> x.getGroups().contains(group)).toArray(TeamInfo[]::new);
    }

    public String[] getHashTags() {
        if (hashtags != null) {
            return hashtags;
        }
        ArrayList<String> hashtags = new ArrayList<>();
        TeamInfo[] infos = getStandings();
        for (TeamInfo teamInfo : infos) {
            if (teamInfo.getHashTag() != null) {
                hashtags.add(teamInfo.getHashTag());
            }
        }
        return hashtags.toArray(new String[0]);
    }

    public abstract long[] firstTimeSolved();

    public abstract RunInfo[] firstSolvedRun();

    public abstract RunInfo[] getRuns();

    public abstract RunInfo getRun(int id);

    public abstract int getLastRunId();
}

File: backend/src/main/kotlin/org/icpclive/events/ContestInfo.kt
package org.icpclive.events

import org.icpclive.api.ContestStatus
import org.icpclive.events.EventsLoader.Companion.instance
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.*
import kotlin.math.min

abstract class ContestInfo{
    abstract val problemsNumber: Int
    abstract val teamsNumber: Int
    var startTime: Long = 0
        set(value) {
            logger.info("Set start time " + Date(startTime))
            field = value
        }
    var problems: MutableList<ProblemInfo> = ArrayList()
    var lastTime: Long = 0
    var status = ContestStatus.BEFORE
        set(value) {
            if (field != value) logger.info("New status: $value")
            lastTime = currentTime
            field = value
        }


    open val timeFromStart: Long
        get() = ((System.currentTimeMillis() - startTime) * instance.emulationSpeed).toLong()
    val currentTime: Long
        get() = getCurrentTime(0)

    private fun getCurrentTime(delay: Int): Long {
        return when (status) {
            ContestStatus.BEFORE -> 0
            ContestStatus.PAUSED -> lastTime
            ContestStatus.RUNNING -> if (startTime == 0L) 0 else min(
                timeFromStart,
                contestLength.toLong()
            )
            ContestStatus.OVER -> {
                if (delay != 0) {
                    min(
                        timeFromStart,
                        (contestLength + delay).toLong()
                    )
                } else contestLength.toLong()
            }
            else -> 0
        }
    }

    val isFrozen: Boolean
        get() = currentTime >= freezeTime

    abstract fun getParticipant(name: String): TeamInfo?
    abstract fun getParticipant(id: Int): TeamInfo?
    abstract fun getParticipantByHashTag(hashTag: String?): TeamInfo?
    abstract val standings: List<TeamInfo>
    abstract fun getStandings(optimismLevel: OptimismLevel): List<TeamInfo>
    fun getStandings(group: String, optimismLevel: OptimismLevel): List<TeamInfo> {
        if (ALL_REGIONS == group) {
            return getStandings(optimismLevel)
        }
        val infos = getStandings(optimismLevel)
        return infos.filter { it.groups.contains(group) }
    }

    abstract fun firstTimeSolved(): LongArray?
    abstract val firstSolvedRun: List<RunInfo?>
    abstract val runs: List<RunInfo>
    var contestLength = 5 * 60 * 60 * 1000
    var freezeTime = 4 * 60 * 60 * 1000
    val groups = mutableSetOf<String>()

    companion object {
        const val ALL_REGIONS = "all"
        val logger: Logger = LoggerFactory.getLogger(ContestInfo::class.java)
    }
}